File: SolaceApp/Services/CloudFunctionAIChatService.swift
import Foundation
import FirebaseFunctions
import SolaceCore

/// Calls the `chatWithAI` Cloud Function, which proxies OpenAI's API.
/// The OpenAI API key lives only in Firebase Functions' secret storage —
/// never on-device and never in this repo.
final class CloudFunctionAIChatService: AIChatServicing {
    private let functions = Functions.functions()

    func sendMessage(history: [AIChatMessage], newMessage: String) async throws -> String {
        let callable = functions.httpsCallable("chatWithAI")
        let payload: [String: Any] = [
            "history": history.map { ["role": $0.role.rawValue, "text": $0.text] },
            "message": newMessage
        ]
        do {
            let result = try await callable.call(payload)
            guard let data = result.data as? [String: Any], let reply = data["reply"] as? String else {
                throw AIChatError.unknown("The assistant returned an unexpected response.")
            }
            return reply
        } catch let error as AIChatError {
            throw error
        } catch {
            throw Self.mapFunctionsError(error)
        }
    }

    private static func mapFunctionsError(_ error: Error) -> AIChatError {
        let nsError = error as NSError
        guard nsError.domain == FunctionsErrorDomain, let code = FunctionsErrorCode(rawValue: nsError.code) else {
            return .unknown(error.localizedDescription)
        }
        switch code {
        case .unauthenticated:
            return .notSignedIn
        case .resourceExhausted:
            return .rateLimited
        case .unavailable, .deadlineExceeded:
            return .network
        default:
            return .unknown(error.localizedDescription)
        }
    }
}

File: SolaceAndroid/app/src/main/java/com/jonathanalumasa/solace/service/firebase/CloudFunctionAIChatService.kt
package com.jonathanalumasa.solace.service.firebase

import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.functions.FirebaseFunctionsException
import com.jonathanalumasa.solace.model.AIChatMessage
import com.jonathanalumasa.solace.service.AIChatError
import com.jonathanalumasa.solace.service.AIChatService
import kotlinx.coroutines.tasks.await

/**
 * Calls the `chatWithAI` Cloud Function, which proxies OpenAI's API — the same
 * function the iOS app uses. The OpenAI key lives only in Firebase Functions'
 * secret storage, never on-device.
 */
class CloudFunctionAIChatService(
    private val functions: FirebaseFunctions = FirebaseFunctions.getInstance()
) : AIChatService {

    override suspend fun sendMessage(history: List<AIChatMessage>, newMessage: String): String {
        val payload = mapOf(
            "history" to history.map { mapOf("role" to it.role.rawValue, "text" to it.text) },
            "message" to newMessage
        )

        val result = try {
            functions.getHttpsCallable("chatWithAI").call(payload).await()
        } catch (error: Exception) {
            throw mapFunctionsError(error)
        }

        @Suppress("UNCHECKED_CAST")
        val data = result.getData() as? Map<String, Any?>
        return data?.get("reply") as? String
            ?: throw AIChatError.Unknown("The assistant returned an unexpected response.")
    }

    private fun mapFunctionsError(error: Exception): AIChatError {
        val functionsError = error as? FirebaseFunctionsException
            ?: return AIChatError.Unknown(error.localizedMessage ?: "Something went wrong.")

        return when (functionsError.code) {
            FirebaseFunctionsException.Code.UNAUTHENTICATED -> AIChatError.NotSignedIn
            FirebaseFunctionsException.Code.RESOURCE_EXHAUSTED -> AIChatError.RateLimited
            FirebaseFunctionsException.Code.UNAVAILABLE,
            FirebaseFunctionsException.Code.DEADLINE_EXCEEDED -> AIChatError.Network
            else -> AIChatError.Unknown(
                functionsError.localizedMessage ?: "Something went wrong."
            )
        }
    }
}
